hal/imx_rt.c

```diff
@@ -876,23 +939,33 @@ int RAMFUNCTION hal_flash_write(uint32_t address, const uint8_t *data, int len)
      * (note 4 p.279 in i.MX RT1060 Processor Reference Manual, Rev. 3, 07/2021)
      */
     asm volatile("cpsid i");
+    bool program_success = true;
     for (i = 0; i < len; i+= CONFIG_FLASH_PAGE_SIZE) {
         memcpy(wbuf, data + i, CONFIG_FLASH_PAGE_SIZE);
         status = g_bootloaderTree->flexSpiNorDriver->program(0, FLEXSPI_CONFIG,
             (address + i) - FLASH_BASE, wbuf);
-        /**
-         * Flash is memory mapped, so the address range must be invalidated in data cache
-         * to ensure coherency between flash and cache
-         */
-        DCACHE_InvalidateByRange(address + i, sizeof(wbuf));
         if (status != kStatus_Success)
         {
-            asm volatile("cpsie i");
-            return -1;
+            program_success = false;
+            break;
         }
     }
+    /* Ensure no speculative prefetching happens before flash program is finished */
+    asm volatile("dsb");
+    /**
+     * Flash is memory mapped, so the address range must be invalidated in data cache
+     * to ensure coherency between flash and cache.
+     * 
+     * Also, both the address and size must be 32-byte aligned as cache-lines are 32 bytes
+     * (see definition of DCACHE_InvalidateByRange).
+     * To ensure all data is included we align the address downwards, and the length upwards.
+     */
+    uint32_t aligned_address = address - (address % 32);
+    uint32_t aligned_len = len + (32 - (len % 32));
+    DCACHE_InvalidateByRange(aligned_address, aligned_len);
+    /* Re-enable interrupts */
     asm volatile("cpsie i");
-    return 0;
+    return program_success;
```

Comment:
Although our return checking for writes is lax, technically this function should return 0 on sucess and -1 on failure, according to the docs. If we could strip the use of `bool` and return these codes that would be great